xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml

```diff
@@ -976,28 +977,193 @@ XWiki.Annotation = Class.create({
+    this.removeAnnotationsAndSelectionSignatures();
+    var annSystem = this;
+    annotations.each(function(ann) {
+      var plainTextStartOffset = ann.fields.find(field =&gt; field.name == 'plainTextStartOffset').value
+      var plainTextEndOffset = ann.fields.find(field =&gt; field.name == 'plainTextEndOffset').value
+      var annDOMRange = annSystem.getDOMRange(annSystem.annotatedElement, plainTextStartOffset, plainTextEndOffset);
+      // Since the annotation could start at a specific offset, the node is split for not wrapping the whole text and a
+      // new range is created to recalculate the new ends.
+      var strictRange = annSystem.fixRangeEnds(annDOMRange);
+
+      // All the nodes inside this range are extracted. Each node it is wrapped inside the annotation marker span.
+      var nodes = annSystem.getTextNodesInRange(strictRange);
+      for(var i = 0; i &lt; nodes.length; i++) {
+        annSystem.markAnnotation(nodes[i], ann);
+      }
+    });
+    this.addAnnotationMarkers(annotations);
+    // Notify the content change.
+    document.fire('xwiki:dom:updated', {elements: [this.annotatedElement]});
+    // Also handle the tab 'downstairs' when the annotations list changes.
+    this.reloadTab(navigateToPane);
     if (andShow) {
       this.toggleAnnotations(true);
     }
   },
 
+  /**
+   * Surround this node with a span corresponding to it's annotation.
+   *
+   * @param markedNode the node that corresponds to the current annotation
+   * @param ann object holding information about the annotation
+  */
+  markAnnotation: function(markedNode, ann) {
+    var wrapper = document.createElement('span');
+    wrapper.addClassName('annotation');
+    wrapper.addClassName('ID' + ann.annotationId);
+
+    var parentNode = markedNode.parentElement;
+    parentNode.replaceChild(wrapper, markedNode);
+    wrapper.appendChild(markedNode);
+  },
+
+  /**
+   * For the first and last node, split the nodes at the known offset to not annotate the whole text. Create a new range
+   * with these new nodes.
+   *
+   * @param range the DOM range of the annotated text
+  */
+  fixRangeEnds: function(range) {
+    var strictRange = new Range();
+
+    var newStartNode = range.startContainer.splitText(range.startOffset);
+    strictRange.setStart(newStartNode, 0);
+
+    var newEndNode = range.endContainer.splitText(range.endOffset);
+    strictRange.setEnd(newEndNode, 0);
+
+    return strictRange;
+  },
+
+  /**
+   * Create a DOM Range by knowing the start and end index from inside the plain content of the element.
+   *
+   * @param annotatedElement the element from where the range is constructed
+   * @param startIndex start offset where the range begins
+   * @param endIndex end offset where the range end
+  */
+  getDOMRange: function(annotatedElement, startIndex, endIndex) {
+    this.annotatedElementPlainText = "";
+    var startPosition = this.getElementAtOffset(annotatedElement, startIndex);
```

Comment:
You don't need to store / keep ``annotatedElementPlainText`` if you remove the length from the ``startIndex`` as you call the method recursively. Same below for ``endIndex``.